README.md
### Build docker image

build:

`docker build . -t signingserver`

run:

 `docker run --rm -d -p 8443:8443 signingserver`

### ETH API

#### 1. 给服务器3个共享密钥中的2个，并提供钱包地址

Request:
```
curl --request POST \
  --url https://localhost:8443/shares \
  --header 'content-type: application/json' \
  --data '{
	"address": "0x040e7783A06e9b994F6e90DF5b2933C03F1b8F21",
 "splits":["80193e13ee645859df5f94935ae2c56525776aa717e105652bf5554a9f58570ade733", "8028d527f1d6a518d7a1deaffae3026d1cb20890d8925a2d5510de7ee7403921bd734"]
}'
```

Response:
```
{
    "address": "0x040e7783A06e9b994F6e90DF5b2933C03F1b8F21",
    "splits": [
        "80193e13ee645859df5f94935ae2c56525776aa717e105652bf5554a9f58570ade733",
        "8028d527f1d6a518d7a1deaffae3026d1cb20890d8925a2d5510de7ee7403921bd734"
    ]
}
```

成功时返回同样的值

#### 2. 根据用户txid查询存款状态

Request:

```
curl --request GET \
  --url https://localhost:8443/deposit/0x5ab5550c6f4bdb32ee0e161dc161d3a4cc45326ee95a5253640c5cac5061bff8?network=ropsten
```

- 查询参数network的值可以为: ropsten, rinkeby, kovan, mainnet 其中之一

Response:

```
{
    "txid": "0x5ab5550c6f4bdb32ee0e161dc161d3a4cc45326ee95a5253640c5cac5061bff8",
    "from": "0x602ea273b80d2116ca4f1178002610aa7e82f8cd",
    "to": "0xcb5a1d293f9c10ea2f996db44fdfe129688a80ba",
    "status": true,
    "includedIn": 3788422,
    "currentHeight": 3794508
}
```


`txid`: 交易号

`from`: 转账方地址

`to`: 接收方地址

`status`: 交易是否成功标识

`includedIn`: 该交易在哪一个块被打包

`currentHeight`: 当前区块高度

可以根据`includedIn`和`currentHeight`差值确定该交易是否足够安全，建议6个区块以上

#### 3. 用户取款，需要提供另外1/3份密钥，打款地址和打款金额(单位:wei)


Request:

```
curl --request POST \
  --url https://localhost:8443/withdraw \
  --header 'content-type: application/json' \
  --data '{
    "from":
    "nonce":
    "gasPrice":
    "gasLimit":
    "split": "8031f1898a9c645ebcf45e5e31e7023fc5c8d2e679ad55c9231495cf066b8fce6eaff",
    "to": "0x14571A8f98301DB5dC5c7640A9C7f6CA5BEaB338",
    "value": 123
}'
```
`from`: 从哪个地方取款

`split`: 被拆分的3份密钥之一

`to`: 接收地址

`value`: 转账数量(单位: wei)

Response:

```
{
    "from":
    "to":
    "rawTx":
    "txid": "0xca9d48cb163f6501a89b0c8b6584691fd9e9820896828f48baf01de9c938e6f5"
}
```

`txid`: 交易号

### BTC API

#### 1. withdraw btc

POST /btc/withdraw

```json
{
	"network": "testnet",
    "split":"8034fd6d9d3b9734e860238cea2975033339e56e6b2841a25a2c028538e9c708900ee",
    "feeUpperLimit": 10000000,
    "fromAddr": "mzJ1AAKQpMj5eaCL3b4oNuSantXmVgz2tM",
    "toAddrs": [
        {
            "address": "n38uMS8K3sM1PfypMd55YH8U4pUrSF4Jqo",
            "amount": 105572000
        },
        {
            "address": "ms8bQ6zx3JWKMa9w7sxZTk1M2ae5AYsbeJ",
            "amount": 7000011
        }

    ]
}

```

* network - testnet or livenet
* split - one of the secret share
* feeUpperLimit - increase the fee below feeUpperLimit will be accepted, otherwise rejected
* fromAddr - from which this transaction is issued
* toAddrs - destination address and corresponding amount

on error, return { "error": "xxxxx" }

```json
{
    "rawTx": "010000000245e3c8d015aa83455f865043ed9b44336cf03397978703922235a8e6ecd4f426010000006a4730440220506746f0a7d619fd8ddce8b32c3df6a2c5354de3d6197e3f72cf3e7e0237e40102204a9a26ba3e20918f4cbc21c29fe7e96c095976aafb1b45d9dac17ee8e722e5ca01210305a491aee8653f88a75534a7f4f0cb34efece7e38d57ff66aba82e42bc257b27fdfffffff0e80b14db20021e5e24c64595ce05c8c45d61231f4abf9bea93fc5d60057af5010000006b483045022100aeeea23af6be7738ad55fe98e7cf84045266ad93fa83ef284fd0e9d07707e5130220281d3a9aadff572807488e7f56595952eba487ddfa89c4132442ccd6a884626b01210305a491aee8653f88a75534a7f4f0cb34efece7e38d57ff66aba82e42bc257b27fdffffff03a0e64a06000000001976a914ed272913a773e3b4288088859bed0ef5f501316688accbcf6a00000000001976a9147f66ea580b5d61169714cc35ecdf27e3f2cbf27788ac95636f00000000001976a914cdf75f817ef312950719f6fa1b947e75ab792d2688ac00000000",
    "txid": "0b7a36ac432161be097ced343c064f12b1ce7bc6b81b1b26c2fa3bf56fabba45",
    "fee": 10000
}
```

* rawTx - raw transaction that can be sent to bitcoin network
* txid - transaction id
* fee - transaction fee

on error, return { "error": "xxxxx" }

#### 2. secret shares

POST /btc/shares

```json
{
    "network": "testnet" | "livenet",
    "address": "mzJ1AAKQpMj5eaCL3b4oNuSantXmVgz2tM",
    "splits": [
        "801722d119769a82fef7660102a3a5341d80fdaf665f8c6ba10bada0a2640c1797b2e",
        "802a630d783f110fdcfcf1ce707c7da6fdb10ff144ed229cde0f9b688eb71670a3a34"
    ]
}
```

#### 3. re-send transaction

POST /btc/re-send

```json
{
    "network": "testnet" | "livenet",
	"originTxid": "3b1f4b8569e532089eaf288afc5f72ca497bb77ab853a759ab5e6119c6e01ca6",
    "split":"8034fd6d9d3b9734e860238cea2975033339e56e6b2841a25a2c028538e9c708900ee",
    "feeUpperLimit": 300000,
    "fromAddr": "mzJ1AAKQpMj5eaCL3b4oNuSantXmVgz2tM",
    "priority": "medium"
}
```

* network - "livenet" | "testnet"
* originTxid - original transaction id that need bump fee
* split - one of the secret share
* priority - "low" | "medium" | "high"
* feeUpperLimit - increase the fee below feeUpperLimit will be accepted, otherwise rejected

```json
{
    "rawTx": "010000000145e3c8d015aa83455f865043ed9b44336cf03397978703922235a8e6ecd4f426010000006b483045022100a5faffb4b66b59eeaa3a3f4b2e58e439ad3dc67fc6cf9571cf31c66fe584fef102207273b39da3d365dcbb15fc80e5c764089858f51cab06c5fba4b5f932ce2e35cf01210305a491aee8653f88a75534a7f4f0cb34efece7e38d57ff66aba82e42bc257b27fdffffff03629c0100000000001976a914ed272913a773e3b4288088859bed0ef5f501316688ac70110100000000001976a9147f66ea580b5d61169714cc35ecdf27e3f2cbf27788ac0e644606000000001976a914cdf75f817ef312950719f6fa1b947e75ab792d2688ac00000000",
    "originTxid": "3b1f4b8569e532089eaf288afc5f72ca497bb77ab853a759ab5e6119c6e01ca6",
    "newTxid": "c029aebfa40dcf8da2b137aafac813e8963c69648bbea80da2dfa1a473f86ad3",
    "fee": 120000
}
```

* rawTx - raw transaction that can be sent to bitcoin network
* originTxid - original transaction id that need bump fee
* newTxid - transaction id after bumping fee
* fee - fee after bumping


CHANGELOG.md
## 2018-11-14

* 比特币支持主网和测试网
* 以太坊支持 `ropsten, rinkeby, kovan, mainnet`网络
* 以太坊 `/deposit` 增加查询参数 `network`, 例如: `https://localhost:8443/deposit/0x5ab5550c6f4bdb32ee0e161dc161d3a4cc45326ee95a5253640c5cac5061bff8?network=ropsten`
* 比特币 `/btc/re-send` 新增 `network` 参数， `network`的值为 `livenet` 和 `testnet` 其中之一

## 2018-11-15

* 支持在多个地址上存取款

test.js
const secrets = require('secrets.js');
const utils = require('./utils');
require('express-async-errors');

const testAddressMatchShares = () => {
    const privateKey = 'abd952e991fb40a146291e6c537fc0db0d1b6de0a815df11efb7e73e1e50daf8'; // must be '0x' prefixed
    const address = '0x040e7783A06e9b994F6e90DF5b2933C03F1b8F21';
    const shares = secrets.share(privateKey, 5, 3);
    console.log(shares)
    console.assert(utils.addressMatchShares(address, shares.slice(1,4)), true);
};

const testBtcAddressMatchShares = () => {
    const privateKey = 'b26412fd29bab29511990b02a1cd20c5d51d17987d4a80f6447293da5fe024e6';
    const address = 'mzJ1AAKQpMj5eaCL3b4oNuSantXmVgz2tM';
    const shares = secrets.share(privateKey, 5, 3);
    console.log(shares);
    console.assert(utils.btcAddressMatchShares(address, "testnet", shares.slice(1, 4)), true);
}

testBtcAddressMatchShares();


const testCoinSeletor = async () => {
    let address = 'mzJ1AAKQpMj5eaCL3b4oNuSantXmVgz2tM';

    console.log(await utils.coinSelector(address, 453001));
}

testCoinSeletor()

const testGetConfirmedUtxos = async () => {
    let a = await utils.getConfirmedUtxo('myxeUAtbtdqWkG1xBwDyB98MV86wcPvLBW')

    console.log(a)
}

testGetConfirmedUtxos()

utils.js
const Web3 = require('web3');
const Tx = require('ethereumjs-tx');
const secrets = require('secrets.js');
const fetch = require("node-fetch");
const bitcore = require('bitcore-lib');

const RopstenWeb3Provider = "https://ropsten.infura.io/Y4zS49bjsYwtRU3Tt4Yj";
const RinkebyWeb3Provider = "https://rinkeby.infura.io/Y4zS49bjsYwtRU3Tt4Yj";
const KovanWeb3Provider = "https://kovan.infura.io/Y4zS49bjsYwtRU3Tt4Yj";
const MainnetWeb3Provider = "https://mainnet.infura.io/Y4zS49bjsYwtRU3Tt4Yj";

const RopstenWeb3 = new Web3(new Web3.providers.HttpProvider(RopstenWeb3Provider));
const RinkebyWeb3 = new Web3(new Web3.providers.HttpProvider(RinkebyWeb3Provider));
const KovanWeb3 = new Web3(new Web3.providers.HttpProvider(KovanWeb3Provider));
const MainnetWeb3 = new Web3(new Web3.providers.HttpProvider(MainnetWeb3Provider));

const BtcLivenet = "http://39.104.129.43:3001";
const BtcTestnet = "http://39.104.129.43:3002"
const MinerFeeOracle = "https://api.blockcypher.com/v1/btc/main";

const addressMatchShares = (address, shares) => {
    const privateKey = secrets.combine(shares);
    const x = MainnetWeb3.eth.accounts.privateKeyToAccount('0x' + privateKey);

    return x.address === address;
};

const reConstructPrivateKey = (shares) => {
    return secrets.combine(shares);
}

const getTransactionReceipt = async (txid, network) => {
    let detail;

    switch (network) {
        case "mainnet":
            detail = await MainnetWeb3.eth.getTransactionReceipt(txid);
            break;
        case "ropsten":
            detail = await RopstenWeb3.eth.getTransactionReceipt(txid);
            break;
        case "rinkeby":
            detail = await RinkebyWeb3.eth.getTransactionReceipt(txid);
            break;
        case "kovan":
            detail = await KovanWeb3.eth.getTransactionReceipt(txid);
            break;
        default:
            // placeholder
            break;
    }

    return detail;
};

const getBlockNumber = async (network) => {
    let blkno;

    switch (network) {
        case "mainnet":
            blkno = await MainnetWeb3.eth.getBlockNumber();
            break;
        case "ropsten":
            blkno = await RopstenWeb3.eth.getBlockNumber();
            break;
        case "rinkeby":
            blkno = await RinkebyWeb3.eth.getBlockNumber();
            break;
        case "kovan":
            blkno = await KovanWeb3.eth.getBlockNumber();
            break;
        default:
            // placeholder
            break;
    }

    return blkno;
}

const buildRawTransaction = async (privateKey, to, nonce, gasPrice, gasLimit, value) => {
    const rawTx = {
        nonce: MainnetWeb3.utils.toHex(nonce),
        gasPrice: MainnetWeb3.utils.toHex(gasPrice),
        gasLimit: MainnetWeb3.utils.toHex(gasLimit),
        to: to,
        value: MainnetWeb3.utils.toHex(value)
    }
    const tx = new Tx(rawTx);
    tx.sign(new Buffer(privateKey, 'hex'));
    const serializedTx = tx.serialize();
    const txid = tx.hash().toString('hex');
    const rawHex = '0x' + serializedTx.toString('hex');

    return [rawHex, txid];
}
// ---------------------btc----------------------

// get utxos that confirmations >= 1
const getConfirmedUtxo = async (address, network) => {
    let endpoint;

    if (network === "livenet") {
        endpoint = BtcLivenet + `/insight-api/addr/${address}/utxo`;
    } else if (network === "testnet") {
        endpoint = BtcTestnet + `/insight-api/addr/${address}/utxo`;
    }

    let response = await fetch(endpoint);
    let utxos = await response.json();

    return utxos.filter(u => u.confirmations >= 1);
}

const getTxInfo = async (txid, network) => {
    let endpoint;

    if (network === "livenet") {
        endpoint = BtcLivenet + `/insight-api/tx/${txid}`;
    } else if (network === "testnet") {
        endpoint = BtcTestnet + `/insight-api/tx/${txid}`;
    }

    let response = await fetch(endpoint);
    if (response.status !== 200) {
        throw new Error("Cant find this transaction")
    }
    let txinfo = await response.json();

    return txinfo;
}

const estimateMinerFee = async () => {
    let response = await fetch(MinerFeeOracle);
    let json = await response.json();

    return {
        "high": json.high_fee_per_kb,
        "medium": json.medium_fee_per_kb,
        "low": json.low_fee_per_kb
    }
}

const getBalance = async (address, network) => {
    let endpoint;

    if (network === "livenet") {
        endpoint = BtcLivenet + `/insight-api/addr/${address}/balance`;
    } else if (network === "testnet") {
        endpoint = BtcTestnet + `/insight-api/addr/${address}/balance`;
    }

    let response = await fetch(endpoint);

    return await response.json();
}

// inspired from: https://zhuanlan.zhihu.com/p/36030990
const coinSelector = async (address, amount, network) => {
    let totalBalance = await getBalance(address, network)
    if (totalBalance < amount) {
        throw new Error("Not enough funds");
    }

    const utxos = await getConfirmedUtxo(address, network);

    // if there is an utxo match `amount`, then return this utxo
    for (var i = 0; i < utxos.length; i++) {
        if (utxos[i].satoshis === amount) {
            return utxos[i];
        }
    }

    // find all utxos that less than amount and check if the sum is equal to amount
    const picked = utxos.filter(u => u.satoshis < amount);
    var sum = 0;
    for (i = 0; i < picked.length; i++) {
        sum += picked[i].satoshis;
    }

    if (sum === amount) {
        return picked;
    }

    // if all those utxo less than amount sumed up less than `amount`,
    // chose the first utxo that greater than `amount`
    if (sum < amount) {
        for (i = 0; i < utxos.length; i++) {
            if (utxos[i].satoshis > amount) {
                return utxos[i];
            }
        }
    }

    // none of the above conditons matched:
    // TODO: sort by mutiple metrics
    utxos.sort((x, y) => x.satoshis < y.satoshis);
    let accumulated = 0;
    let result = [];
    for (i = 0; i < utxos.length; i++) {
        accumulated += utxos[i].satoshis;
        result.push(utxos[i]);
        if (accumulated > amount) {
            return result;
        }
    }
}

const btcAddressMatchShares = (address, network, shares) => {
    let privateKey = secrets.combine(shares);
    privateKey = new bitcore.PrivateKey(privateKey);
    const publicKey = privateKey.toPublicKey();

    if (network === "testnet") {
        var addr = publicKey.toAddress(bitcore.Networks.testnet).toString();
    } else if (network === "livenet") {
        var addr = publicKey.toAddress(bitcore.Networks.livenet).toString();
    } else {
        throw new Error("unsupported network");
    }

    return addr === address;
}

const BtcTx = () => {
    return new bitcore.Transaction();
}

module.exports = {
    addressMatchShares: addressMatchShares,
    getTransactionReceipt: getTransactionReceipt,
    getBlockNumber: getBlockNumber,
    reConstructPrivateKey: reConstructPrivateKey,
    buildRawTransaction: buildRawTransaction,

    // ------ btc --------
    getConfirmedUtxo: getConfirmedUtxo,
    coinSelector: coinSelector,
    getBalance: getBalance,
    btcAddressMatchShares: btcAddressMatchShares,
    BtcTx: BtcTx,
    getTxInfo: getTxInfo,
    estimateMinerFee: estimateMinerFee
};

app.js
const express = require('express');
require('express-async-errors');
const bodyParser = require('body-parser');

const utils = require('./utils');
const bitcore = require('bitcore-lib');

const https = require("https");
const fs = require("fs");

const EthAddresses = new Map();
const BtcAddresses = new Map();

const app = express();
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());

app.post('/shares', (req, res) => {
    const splits = req.body.splits;
    const address = req.body.address;

    if(!(splits && address && splits.length === 2)) {
        res.status(400).json({
            "error": "bad request"
        });
    }

    EthAddresses.set(address, {splits: splits})

    res.status(200).json(EthAddresses.get(address));
});

app.post('/withdraw', async (req, res) => {
    const from = req.body.from;
    const split = req.body.split;
    const nonce = req.body.nonce;
    const gasPrice = req.body.gasPrice;
    const gasLimit = req.body.gasLimit;
    const to = req.body.to;
    const value = req.body.value;
    let tx;

    const tmp = EthAddresses.get(from).splits.concat(split);

    if(utils.addressMatchShares(from, tmp)) {
        const privateKey = utils.reConstructPrivateKey(tmp);
        tx = await utils.buildRawTransaction(privateKey, to, nonce, gasPrice, gasLimit, value);
        res.status(200).json({
            "from": from,
            "to": to,
            "rawTx": tx[0],
            "txid": "0x" + tx[1]
        });
    } else {
        res.status(400).json({
            "error": "bad request"
        });
    }
});

app.get('/deposit/:txid', async (req, res) => {
    const txid = req.params.txid;
    const network = req.query.network;

    try {
        const receipt = await utils.getTransactionReceipt(txid, network);
        const blkNo = await utils.getBlockNumber(network);

        res.status(200).json({
            "txid": txid,
            "from": receipt.from,
            "to": receipt.to,
            "status": receipt.status,
            "includedIn": receipt.blockNumber,
            "currentHeight": blkNo,
        });
    } catch(e) {
        res.status(400).json({
            "error":"network error"
        });
        console.log(e)
    }
});

// --------------- btc -----------------

app.post('/btc/shares', async (req, res) => {
    const splits = req.body.splits;
    const address = req.body.address;
    const network = req.body.network;

    if(!(splits && address && splits.length === 2)) {
        res.status(400).json({
            "error": "parameter error"
        });
    }

    BtcAddresses.set(address, {splits: splits, network: network});

    res.status(200).json(BtcAddresses.get(address));
})

app.post('/btc/withdraw', async (req, res) => {
    const split = req.body.split;
    const feeUpperLimit = req.body.feeUpperLimit;
    const toAddrs = req.body.toAddrs;
    const fromAddr = req.body.fromAddr;
    const network = req.body.network;

    let totalAmount = 0;
    for(var i = 0; i < toAddrs.length; i++) {
        totalAmount += toAddrs[i].amount;
    }

    const totalBalance = await utils.getBalance(fromAddr, network);

    if (totalBalance < totalAmount) {
        res.status(400).json({
            "error": "not enough funds"
        });
        return;
    }

    let minerFee = await utils.estimateMinerFee();
    let fee = minerFee.low;

    const selectedUtxos = await utils.coinSelector(fromAddr, totalAmount + fee, network);

    if (!selectedUtxos) {
        res.status(400).json({
            "error": "no appropriate utxo found"
        });
        return;
    }
    const privateKey = utils.reConstructPrivateKey(BtcAddresses.get(fromAddr).splits.concat(split));
    const shares = BtcAddresses.get(fromAddr).splits.concat(split);

    try {
        if(!utils.btcAddressMatchShares(fromAddr, network, shares)) {
            res.status(400).json({
                "error": "Address not match the shares"
            });
            return;
        }
    } catch(e) {
        res.status(400).json({
            "error": "Can't re-construct secret from shares"
        });
        console.log(e);
        return;
    }

    const tx = utils.BtcTx();

    if (Array.isArray(toAddrs)) {
        for (i = 0; i < toAddrs.length; i++) {
            tx.to(toAddrs[i].address, toAddrs[i].amount);
        }
    } else {
        tx.to(toAddrs.address, toAddrs.amount);
    }


    try {
        tx.from(selectedUtxos)
            .change(fromAddr)
            .enableRBF()
            .feePerKb(fee)
            .sign(privateKey);
    } catch(e) {
        console.log(selectedUtxos)
        console.log(e)
        res.status(400).json({
            "error": "Incorrect signing key"
        });
        return;
    }

    if (fee > feeUpperLimit) {
        res.status(400).json({
            "error": "fee reach upper limit"
        });
    } else {
        try {
            const rawTx = tx.serialize();
            const txid = tx.hash;
            res.status(200).json({
                "rawTx": rawTx,
                "txid": txid,
                "fee": fee
            });
        } catch(e) {
            console.log(e)
            res.status(400).json({
                "error": "Can't serialize raw transaction"
            });
            console.log(e)
        }
    }
})

app.post('/btc/re-send', async (req, res) => {
    const originTxid = req.body.originTxid;
    const split = req.body.split;
    const priority = req.body.priority;
    const feeUpperLimit = req.body.feeUpperLimit;
    const network = req.body.network;

    try {
        var txinfo = await utils.getTxInfo(originTxid, network);
    } catch(e) {
        res.status(400).json({
            "error": "re-send an unknow transaction"
        });
        console.log(e);
        return;
    }

    if (txinfo.confirmations > 0 && txinfo.blockheight !== -1) {
        res.status(400).json({
            "error": "Transaction has been succeed"
        });
        return;
    }

    let minerFee = await utils.estimateMinerFee();
    let fee = 0;

    if (priority === "low") {
        fee = minerFee.low;
    } else if (priority === "medium") {
        fee = minerFee.medium;
    } else if (priority === "high") {
        fee = minerFee.high;
    } else {
        res.status(400).json({
            "error": "Please check fee priority"
        });
        return;
    }

    const vin = txinfo.vin;
    const vout = txinfo.vout;
    const utxos = [];
    let fromAddr = "";

    for (var i = 0; i < vin.length; i++) {
        const id = vin[i].txid;
        const vout = vin[i].vout;
        const address = vin[i].addr;
        fromAddr = address; // FIXME
        const satoshis = vin[i].valueSat;

        const addr = bitcore.Address.fromString(address);
        const script = bitcore.Script.buildPublicKeyHashOut(addr);
        const scritpPubkey = script.toHex();

        const utxo = new bitcore.Transaction.UnspentOutput({
            "txid" : id,
            "vout" : vout,
            "address" : address,
            "scriptPubKey" : scritpPubkey,
            "satoshis" : satoshis
        });

        utxos.push(utxo);
    }
    const tx = utils.BtcTx();

    try {
        var privateKey = utils.reConstructPrivateKey(BtcAddresses.get(fromAddr).splits.concat(split));
    } catch(e) {
        res.status(400).json({
            "error": "Cant re-contruct private key"
        });
        console.log(e)
        return;
    }

    for (i = 0; i < vout.length; i++) {
        if (vout[i].scriptPubKey.addresses[0] !== fromAddr) {
            const value = bitcore.Unit.fromBTC(vout[i].value).toSatoshis();
            const address = vout[i].scriptPubKey.addresses[0];
            tx.to(address, value);
        }
    }

    try {
        tx.from(utxos)
        .change(fromAddr)
        .enableRBF()
        .feePerKb(fee)
        .sign(privateKey);
    } catch(e) {
        res.status(400).json({
            "error": "Incorrect siging key"
        });
        console.log(e);
        return;
    }


    if (tx.getFee() > feeUpperLimit) {
        res.status(400).json({
            "error": "fee reach upper limit"
        });
        return;
    }

    res.status(200).json({
        "rawTx": tx.serialize(),
        "originTxid": originTxid,
        "newTxid": tx.hash,
        "fee": tx.getFee()
    })
})

https.createServer({
    key: fs.readFileSync('./keys/key.pem'),
    cert: fs.readFileSync('./keys/cert.pem')
}, app).listen(8443);
